domain/machine/state/state.go

```diff
@@ -0,0 +1,173 @@
+		}
+		// For now, we just care if the minimal machine row already exists.
+		if err == nil {
+			return nil
+		}
+		nodeUUID, err := utils.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		machineUUID, err := utils.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		createParams := sqlair.M{
+			"machine_uuid":  machineUUID.String(),
+			"net_node_uuid": nodeUUID.String(),
+			"machine_id":    machineId,
+			"life_id":       life.Alive,
+		}
+		if err := tx.Query(ctx, createNodeStmt, createParams).Run(); err != nil {
+			return errors.Annotatef(err, "creating net node row for machine %q", machineId)
+		}
+		if err := tx.Query(ctx, createMachineStmt, createParams).Run(); err != nil {
+			return errors.Annotatef(err, "creating machine row for machine %q", machineId)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("upserted machine %q", machineId)
+
+	return nil
+}
+
+// DeleteMachine deletes the specified machine and any dependent child records.
+// TODO - this just deals with child block devices for now.
+func (s *State) DeleteMachine(ctx context.Context, machineId string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	machineIDParam := sqlair.M{"machine_id": machineId}
+
+	queryMachine := `SELECT &M.uuid FROM machine WHERE machine_id = $M.machine_id`
+	queryMachineStmt, err := sqlair.Prepare(queryMachine, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteMachine := `DELETE FROM machine WHERE machine_id = $M.machine_id`
+	deleteMachineStmt, err := sqlair.Prepare(deleteMachine, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteNode := `
+DELETE FROM net_node WHERE uuid IN
+(SELECT net_node_uuid FROM machine WHERE machine_id = $M.machine_id) 
+`
+	deleteNodeStmt, err := sqlair.Prepare(deleteNode, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		result := sqlair.M{}
+		err = tx.Query(ctx, queryMachineStmt, machineIDParam).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "looking up UUID for machine %q", machineId)
+		}
+		// Machine already deleted is a no op.
+		if len(result) == 0 {
+			return nil
+		}
+		machineUUID := result["uuid"].(string)
+
+		if err := blockdevice.RemoveMachineBlockDevices(ctx, tx, machineUUID); err != nil {
+			return errors.Annotatef(err, "deleting block devices for machine %q", machineId)
+		}
+
+		if err := tx.Query(ctx, deleteMachineStmt, machineIDParam).Run(); err != nil {
+			return errors.Annotatef(err, "deleting machine %q", machineId)
+		}
+		if err := tx.Query(ctx, deleteNodeStmt, machineIDParam).Run(); err != nil {
+			return errors.Annotatef(err, "deleting net node for machine  %q", machineId)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("deleted machine %q", machineId)
```

Comment:
Does this really need to exist and if so it should be coming from above the services layer. The fact the stack hasn't errored is the exact same meaning.

Reply:
Yeah, was being consistent with what was done elsewhere. Can remove.